crates/biome_js_analyze/src/analyzers/complexity/no_excessive_cognitive_complexity.rs

```diff
@@ -402,37 +403,50 @@ impl ComplexityOptions {
 }
 
-impl VisitNode<JsonLanguage> for ComplexityOptions {
-    fn visit_map(
-        &mut self,
-        key: &JsonSyntaxNode,
-        value: &JsonSyntaxNode,
+impl Deserializable for ComplexityOptions {
+    fn deserialize<L: DeserializableLanguage>(
+        value: SyntaxNode<L>,
         diagnostics: &mut Vec<DeserializationDiagnostic>,
-    ) -> Option<()> {
-        let (name, value) = self.get_key_and_value(key, value)?;
-        let name_text = name.inner_string_text().ok()?;
-        let name_text = name_text.text();
-        if name_text == "maxAllowedComplexity" {
-            if let Some(value) = value
-                .as_json_number_value()
-                .and_then(|number_value| u8::from_str(&number_value.text()).ok())
-                // Don't allow 0 or no code would pass.
-                // And don't allow MAX_SCORE or we can't detect exceeding it.
-                .filter(|&number| number > 0 && number < MAX_SCORE)
-            {
-                self.max_allowed_complexity = value;
-            } else {
-                diagnostics.push(
-                    DeserializationDiagnostic::new(markup! {
-                        "The field "<Emphasis>"maxAllowedComplexity"</Emphasis>
-                        " must contain an integer between 1 and "{MAX_SCORE - 1}
-                    })
-                    .with_range(value.range()),
-                );
+    ) -> Option<Self> {
+        L::deserialize(value, ComplexityOptionsVisitor, diagnostics)
+    }
+}
+
+struct ComplexityOptionsVisitor;
+impl DeserializationVisitor for ComplexityOptionsVisitor {
+    type Value = ComplexityOptions;
+
+    fn visit_map<L: DeserializableLanguage>(
+        self,
+        members: impl Iterator<Item = (SyntaxNode<L>, SyntaxNode<L>)>,
+        _range: TextRange,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<Self::Value> {
+        let mut result = Self::Value::default();
+        for (key, val) in members {
+            let key_range = key.text_trimmed_range();
+            let Some(key_text) = TokenText::deserialize(key, diagnostics) else {
+                continue;
+            };
+            match key_text.text() {
+                "maxAllowedComplexity" => {
+                    if let Some(val) = Deserializable::deserialize(val, diagnostics) {
+                        result.max_allowed_complexity = val;
+                    }
+                }
+                _ => diagnostics.push(DeserializationDiagnostic::new_unknown_key(
+                    key_text.text(),
+                    key_range,
+                    Self::Value::ALLOWED_KEYS,
+                )),
             }
-        } else {
-            report_unknown_map_key(&name, Self::ALLOWED_KEYS, diagnostics);
         }
-        Some(())
+        Some(result)
+    }
+}
+
+impl Display for ComplexityOptionsVisitor {
+    fn fmt(&self, fmt: &mut Formatter) -> std::io::Result<()> {
+        write!(fmt, "object")
```

Comment:
I am conflicted by the implementation of `Display`, while I understand why it's needed, the reason *why* it needs to be implemented is unclear.

Can't we have a new trait called `ExpectedType` that has a function called `expect_type` that returns a `&'static str` and then pass it to the diagnostic? (instead of passing `self`).

Curious to know what are the constraints

Reply:
We can same add `expect_type` on the `DeserializationVisitor` trait. Using `Display` has the advantage to be closer to serde and allow custom message based on the visitor fields. Maybe it is overkill since we always use a simple static string...

Reply:
Yeah I think it's an overkill and difficult to justify.

 Also, we are in control of the errors we emit, so we have more freedom